### DISCUSSION TITLE
dynamic link creation model

### DISCUSSION THREAD

**Discussion by dragoncoder047** [pschatzmann/arduino-audio-tools#823] [Q&A] [answered]
How is this library modeled to create links between different nodes? There seem to be multiple different base classes for things (AudioStream, AudioOutput, AudioSource, StreamCopy, etc) and none inherit from any common superclass.

The Javascript web audio API, as well as PJRC's Teensy Audio library, designed everything as audio nodes that have input and output "ports", that are then connected together. A sine wave generator might have 1 output (mono) and 0 inputs, for example. I find this kind of organization easy to think about, but unfortunately I'm trying to do some audio work on an ESP32, using this library.

I'm not asking for any rewrite here. I'm just asking is there any base model/abstraction/idea about how audio is routed between nodes? Mostly because I want to dynamically re-route audio at runtime, and knowing when to pass in an input stream as a parameter to a constructor and then using that stream as an output, versus using a StreamCopy to pump data between two streams is a little confusing.

**Comment by pschatzmann** [comment]
The starting point was the Arduino class model! At the root you have Print and as subclass of Print you have Stream. Therefore AudioStream inherits from Stream and AudioOutput inherits from Print!
The common class of all is therefore the Arduino Print.

Arduino Print is for writing data and Stream supports writing and reading! I tried to be consistent in the classes, so when you can read and write it is a Stream and if it is output only it is ending with Output and inherits from Print and not Stream.

This has the big advantage that we can also use any Arduino subclasses which inherits from Print of Stream: in our framwork as well: e.g. File, Serial etc...

**Reply by dragoncoder047** [reply]
What about AudioSource and StreamCopy? Neither of them list any superclasses in the class definition, they don't inherit from Print. **What are the uses for them?**

**Reply by pschatzmann** [reply]
I created AudioSource as abstract data api for the audio player, that supports forward and backward navigation: it just navigation logic that provides a Stream as result to be played! Most effective implementations are then basically just for the different Arduino File APIs

StreamCopy is just a utility class that copies data from a source to a destination. The source obiously needs to be inherited from Stream (which provides reading) and the destination can be both a Stream or Print/Output class because both implement writing.

**Reply by pschatzmann** [reply]
Here is a link to the [Print documentation](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_print.html)

**Reply by dragoncoder047** [reply]
So correct me if I'm wrong:

* AudioSource is strictly for encoded streams like MP3 of AAC files off an SD card, the AudioSourceSD is passed into an EncodedAudioStream along with the required codec for decoding it.
* StreamCopy is only used when you don't have a stream at the end that implements .loop() to read its inputs and update its output buffers so you need to manually push data into it using the copier.
* everything else is either passed the input in the constructor (if only one input) or using .add() for example the MixerNode, and you call .loop() on the root node to read everything up the tree and update outputs.

**Reply by pschatzmann** [reply]
I corrected the explanation for the AudioSource!

**Reply by pschatzmann** [reply]
StreamCopy is used in an Arduino loop() to push the data from a source to a destination. You could do reads or writes yourself as explained in the Introduction Wiki, but this saves you from some pitfalls and is much less code to write

**Reply by dragoncoder047** [reply]
> I corrected the explanation for the AudioSource!

Well, then where do the codecs come in? You can't just play an MP3 as raw data, you'll just get noise.

Also I noticed according to the documentation, it seems odd that if you have a Stream and want to add an effect to it, you have to pipe it through a GeneratorFromStream because AudioEffects will only take a AudioGenerator as input. (Also, you've marked AudioEffects as obsolete. What's the replacement?)

**Reply by pschatzmann** [reply, marked as answer]
Just read the chapter about encoding and decoding
The Wiki is quite complete!

**Reply by dragoncoder047** [reply]
Thank you. I didn't know the wiki existed, and the doxygen documentation is a little cryptic.
